xdsl/transforms/memref_to_dsd.py

```diff
@@ -0,0 +1,203 @@
+        zeros_op = csl.ZerosOp(op.memref.type)
+
+        dsd_t = csl.DsdType(
+            csl.DsdKind.mem1d_dsd
+            if len(op.memref.type.shape) == 1
+            else csl.DsdKind.mem4d_dsd
+        )
+        offsets = None
+        if isinstance(op.memref.type.layout, StridedLayoutAttr) and isinstance(
+            op.memref.type.layout.offset, IntAttr
+        ):
+            offsets = ArrayAttr([IntegerAttr(op.memref.type.layout.offset, 16)])
+
+        shape = [arith.Constant(IntegerAttr(d, 16)) for d in op.memref.type.shape]
+        dsd_op = csl.GetMemDsdOp.build(
+            operands=[zeros_op, shape],
+            result_types=[dsd_t],
+            properties={
+                "offsets": offsets,
+            },
+        )
+
+        rewriter.replace_matched_op([zeros_op, *shape, dsd_op])
+
+
+class LowerSubviewOpPass(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: memref.Subview, rewriter: PatternRewriter, /):
+        assert isa(op.source.type, MemRefType[Attribute])
+        assert len(op.static_sizes.data) == 1, "not implemented"
+        assert len(op.static_offsets.data) == 1, "not implemented"
+        assert len(op.static_strides.data) == 1, "not implemented"
+        new_ops: list[Operation] = []
+        curr_op = op.source
+
+        # update sizes only if they differ from op.source.type
+        if op.static_sizes.data.data[0].data == memref.Subview.DYNAMIC_INDEX:
+            pass  # todo
+        elif op.static_sizes.as_tuple() != op.source.type.get_shape():
+            new_ops.append(
+                len_op := arith.Constant(
+                    IntegerAttr(
+                        cast(ArrayAttr[IntAttr], op.static_sizes.data).data[0],
+                        csl.u16_value,
+                    )
+                )
+            )
+            new_ops.append(
+                curr_op := csl.SetDsdLengthOp.build(
+                    operands=[curr_op, len_op], result_types=[op.source.type]
+                )
+            )
+
+        # update strides only if they differ from op.source.type
+        if op.static_strides.data.data[0].data == memref.Subview.DYNAMIC_INDEX:
+            pass  # todo
+        elif op.static_strides.as_tuple() != op.source.type.get_strides():
+            new_ops.append(
+                stride_op := arith.Constant(
+                    IntegerAttr(
+                        cast(ArrayAttr[IntAttr], op.static_strides.data).data[0],
+                        IntegerType(8, Signedness.SIGNED),
+                    )
+                )
+            )
+            new_ops.append(
+                curr_op := csl.SetDsdStrideOp.build(
+                    operands=[curr_op, stride_op], result_types=[op.source.type]
+                )
+            )
+
+        # update offsets only if they differ from op.source.type
+        if op.static_offsets.data.data[0].data == memref.Subview.DYNAMIC_INDEX:
+            pass  # todo
```

Comment:
🤔

Reply:
What happens if we hit this? What is to be done?

Reply:
basically the same as for static, working on this now

Reply:
I came to think, if we have a dynamic value, let's assume it is not as de-facto static as coming from an arith.constant (in this case, we should find or write a pass to promote arith.constant to static values). It's probably simpler and cleaner to not analyse dynamic values here and simply generate dsd modify ops for each dynamic value, without analysis.